### PR TITLE
filter() and slice() preserve order across groups

### PR DESCRIPTION
**Pull request by romainfrancois**
I also made `slice()` preserve the original order

``` r
library(dplyr, warn.conflicts = FALSE)
tb <- tibble(g = c(1, 2, 1, 2, 1), time = 5:1, x = 5:1)

tb %>%
  group_by(g) %>%
  arrange(time) %>%
  filter(x <= 4)
#> # A tibble: 4 x 3
#> # Groups:   g [2]
#>       g  time     x
#>   <dbl> <int> <int>
#> 1     1     1     1
#> 2     2     2     2
#> 3     1     3     3
#> 4     2     4     4

tb %>%
  group_by(g) %>%
  filter(x <= 4) %>%
  arrange(time)
#> # A tibble: 4 x 3
#> # Groups:   g [2]
#>       g  time     x
#>   <dbl> <int> <int>
#> 1     1     1     1
#> 2     2     2     2
#> 3     1     3     3
#> 4     2     4     4

tb %>%
  group_by(g) %>%
  arrange(time) %>%
  slice(1,n())
#> # A tibble: 4 x 3
#> # Groups:   g [2]
#>       g  time     x
#>   <dbl> <int> <int>
#> 1     1     1     1
#> 2     2     2     2
#> 3     2     4     4
#> 4     1     5     5

tb %>%
  group_by(g) %>%
  slice(1, n()) %>%
  arrange(time)
#> # A tibble: 4 x 3
#> # Groups:   g [2]
#>       g  time     x
#>   <dbl> <int> <int>
#> 1     1     1     1
#> 2     2     2     2
#> 3     2     4     4
#> 4     1     5     5
```

<sup>Created on 2018-12-11 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

### CODE REVIEW

**Comment by romainfrancois** [comment]
For `slice()` I'm not sure because e.g. `slice(1,n())` and `slice(n(), 1)` give the same result

``` r
library(dplyr, warn.conflicts = FALSE)
tb <- tibble(g = c(1, 2, 1, 2, 1), time = 5:1, x = 5:1)

tb %>% 
  slice(1, n())
#> # A tibble: 2 x 3
#>       g  time     x
#>   <dbl> <int> <int>
#> 1     1     5     5
#> 2     1     1     1

tb %>% 
  slice(n(), 1)
#> # A tibble: 2 x 3
#>       g  time     x
#>   <dbl> <int> <int>
#> 1     1     5     5
#> 2     1     1     1
```

<sup>Created on 2018-12-11 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1.9000)</sup>

**Comment by hadley** [comment]
Hmmm, I think slice might be different — I think people would expect `slice(df, 2, 1)` to reorder.

**Comment by romainfrancois** [comment]
makes sense. I was already too far when I realized it. I'll revert `slice()`

File: src/filter.cpp [hunk 1]
```diff
@@ -32,111 +32,82 @@ SEXP check_result_lgl_type(SEXP tmp) {
...
-      int idx = g_idx[j];
-      if (pred(idx)) {
-        test[-idx - 1] = FALSE;
-      }
+    int ng = idx.size();
+    const int* p_test = g_test.begin();
+
+    int new_size = 0;
+    for (int j = 0; j < ng; j++, ++p_test) {
+      new_size += *p_test == TRUE;
```
Comment (krlmlr) [inline src/filter.cpp]: Can we define a consistent way for testing the trueness of a logical (maybe later)? This code uses `== TRUE`, code below uses `== 1` -- an internal inline function would disambiguate.

File: src/filter.cpp [hunk 2]
```diff
@@ -145,103 +116,61 @@ class GroupFilterIndices {
...
-          }
-        }
-      }
+  // once this has been trained on all groups
+  // this materialize indices and rows
+  void process() {
+    indices = IntegerVector(no_init(k));
+    std::vector<int*> p_rows(ngroups);
+    for (int i = 0; i < ngroups; i++) {
+      SEXP idx = rows[i] = Rf_allocVector(INTSXP, new_sizes[i]);
```
Comment (krlmlr) [inline src/filter.cpp]: The fact that `rows[i]` protects is somewhat obscured here. I'm also a bit torn on the use of multiple assignments in one statement.

File: src/filter.cpp [hunk 3]
```diff
@@ -396,6 +358,119 @@ class CountIndices {
   int n_neg;
 };
 
+template <typename SlicedTibble>
```
Comment (krlmlr) [inline src/filter.cpp]: I wonder if code related to `slice()` should get its own `.cpp` file.

**Pull request merged**